client/kafka/src/main/java/org/hawkular/apm/client/kafka/AbstractPublisherKafka.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.hawkular.apm.api.services.Publisher;
+import org.hawkular.apm.api.services.PublisherMetricHandler;
+import org.hawkular.apm.api.services.ServiceStatus;
+import org.hawkular.apm.api.utils.PropertyUtil;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * This class represents an abstract Kafka producer.
+ *
+ * @author gbrown
+ *
+ * @param <T> The event type being published.
+ */
+public abstract class AbstractPublisherKafka<T> implements Publisher<T>, ServiceStatus {
+
+    private static ObjectMapper mapper = new ObjectMapper();
+
+    private Producer<String, String> producer;
+
+    private String topic;
+
+    private PublisherMetricHandler<T> handler = null;
+
+    /**
+     * This constructor initialises the topic.
+     *
+     * @param topic The topic
+     */
+    public AbstractPublisherKafka(String topic) {
+        this.topic = topic;
+        if (isAvailable()) {
+            init();
+        }
+    }
+
+    /* (non-Javadoc)
+     * @see org.hawkular.apm.api.services.ServiceStatus#isAvailable()
+     */
+    @Override
+    public boolean isAvailable() {
+        return PropertyUtil.hasProperty(PropertyUtil.HAWKULAR_APM_KAFKA_SERVERS);
+    }
+
+    /**
+     * This method initialises the publisher.
+     */
+    protected void init() {
+        Properties props = new Properties();
+        props.put("bootstrap.servers", PropertyUtil.getProperty(PropertyUtil.HAWKULAR_APM_KAFKA_SERVERS));
+        props.put("acks", "all");
+        props.put("retries", PropertyUtil.getPropertyAsInteger(PropertyUtil.HAWKULAR_APM_KAFKA_PRODUCER_RETRIES, 3));
+        props.put("batch.size", 16384);
+        props.put("linger.ms", 1);
+        props.put("buffer.memory", 33554432);
+        props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
+        props.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");
+
+        producer = new KafkaProducer<>(props);
+    }
+
+    /* (non-Javadoc)
+     * @see org.hawkular.apm.api.services.Publisher#getInitialRetryCount()
+     */
+    @Override
+    public int getInitialRetryCount() {
+        return 0;
+    }
+
+    /* (non-Javadoc)
+     * @see org.hawkular.apm.api.services.Publisher#publish(java.lang.String, java.util.List)
+     */
+    @Override
+    public void publish(String tenantId, List<T> items) throws Exception {
+        publish(tenantId, items, getInitialRetryCount(), 0);
+    }
+
+    /* (non-Javadoc)
+     * @see org.hawkular.apm.api.services.Publisher#publish(java.lang.String, java.util.List, int, long)
+     */
+    @Override
+    public void publish(String tenantId, List<T> items, int retryCount, long delay) throws Exception {
+        long startTime=0;
+
+        // Check if delay is required
+        // TODO: May need to check if delay is excessive and schedule message publish in separate thread
+        if (delay > 0) {
```

Comment:
The delay is generally used to provide a time window for other async events to be processed, so that when the delayed event is published, all of the information will be available.